api/v1beta1/keystoneendpoint.go

```diff
@@ -225,19 +227,119 @@ func GetKeystoneEndpointUrls(
 		return nil, err
 	}
 
+	endpointurls := filterVisibility(ke.Items, visibility)
+
+	return endpointurls, nil
+}
+
+// GetKeystoneEndpointUrlsForServices returns all URLs currently registered. Visibility can be admin, public or internal.
+// If it is nil, all URLs are returned. Optional services parameter can be provided to only return the values for
+// those services matching the `service` (common.AppSelector) label key.
+func GetKeystoneEndpointUrlsForServices(
+	ctx context.Context,
+	h *helper.Helper,
+	namespace string,
+	visibility *string,
+	servicesToKeep []string,
+) ([]string, error) {
+
+	ke, err := GetKeystoneEndpointList(ctx, h, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	keystoneEndpoints := ke.Items
+	if len(servicesToKeep) > 0 && ke != nil {
+		keystoneEndpoints = filterServices(keystoneEndpoints, servicesToKeep)
+	}
+
+	endpointurls := filterVisibility(keystoneEndpoints, visibility)
+
+	return endpointurls, nil
+}
+
+// GetHashforKeystoneEndpointUrlsForServices returns a hash for a list of endpoint URLs currently registered.
+// Visibility can be admin, public or internal. If it is nil, all URLs are returned.
+// Optional services parameter can be provided to only return the values for those services matching the
+// `service` (common.AppSelector) label key.
+func GetHashforKeystoneEndpointUrlsForServices(
```

Comment:
I'm wondering if we will hash at the right time. The URL is taken from a Spec not from a Status so the change to switch to https is requested but might not been applied yet, so triggering a restart at this time is racy.

Reply:
the endpoint id at least i think is in the status it may make sense to put the full url there also.

can observed generation help here or is that unrelated.

Reply:
updated the PR to enhance the status of a KeystoneEndpoint to reflect the endpoint details after they get created/updated and using this to get the endpoint urls.